airbyte-integrations/connectors/source-blob/source_blob/stream.py

```diff
@@ -0,0 +1,376 @@
+#
+# Copyright (c) 2020 Airbyte
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from datetime import datetime
+import json
+import concurrent
+from operator import itemgetter
+from typing import Any, Iterable, Iterator, List, Mapping, MutableMapping, Optional, Tuple, Union
+from traceback import format_exc
+from airbyte_cdk.models.airbyte_protocol import SyncMode
+from fnmatch import fnmatch
+from smart_open.s3 import _list_bucket
+from jsonschema import Draft4Validator, SchemaError
+from .fileclient import FileClient, FileClientS3
+from .filereader import FileReaderCsv
+
+from airbyte_cdk.logger import AirbyteLogger
+from airbyte_cdk.sources.streams import Stream
+
+
+class ConfigurationError(Exception):
+    """Client mis-configured"""
+
+
+class FileStream(Stream, ABC):
+    """
+    TODO docstring
+    """
+    format_filereader_map = {
+        'csv': FileReaderCsv,
+        # 'parquet': FileReaderParquet,
+        # etc.
+    }
+    ab_additional_col = "_airbyte_additional_properties"
+    ab_last_mod_col = "_airbyte_source_file_last_modified"
+    ab_file_name_col = "_airbyte_source_file_url"
+    datetime_format_string = "%Y-%m-%dT%H:%M:%S%z"
+
+    def __init__(self, dataset_name: str, provider: dict, format: dict, path_patterns: List[str], schema: str = None):
+        self.dataset_name = dataset_name
+        self._path_patterns = path_patterns
+        self._provider = provider
+        self._format = format
+        self._schema = {}
+        if schema:
+            try:
+                self._schema = json.loads(schema)
+            except json.decoder.JSONDecodeError as err:
+                error_msg = f"Failed to parse schema {repr(err)}\n{schema}\n{format_exc()}"
+                self.logger.error(error_msg)
+                raise ConfigurationError(error_msg) from err
+            try:
+                Draft4Validator.check_schema(self._schema)
```

Comment:
Is this validation too stringent? I think users will just define a key value map (based on the schema example in `spec.json`), which does not follow the format of an official JsonSchema. For example, the JsonSchema object will start with `"type": "object"` at the root level, and define all fields inside `properties`, but users won't do that.

Nit. If the validation from `jsonschema` is not too stringent, should we use `Draft7Validator` instead of `Draft4Validator`?

Reply:
Yes agreed, this was a leftover from original idea to use full JsonSchema syntax before I simplified it to a key:value map of columnName:dataType. Will update the validation accordingly